File: python_files/raylib/single_file_programs/bouncing_balls_v2.py
import math
import random

import pyray as p

SCREEN_WIDTH: int = 1280
SCREEN_HEIGHT: int = 720
MAX_BALLS: int = 400


class Ball:
    def __init__(self) -> None:
        self.radius: float = random.uniform(5, 40)
        self.center: p.Vector2 = p.Vector2(
            random.uniform(self.radius, p.get_screen_width() - self.radius),
            random.uniform(self.radius, p.get_screen_height() - self.radius),
        )
        speed: float = random.uniform(100, 300)
        angle: float = random.uniform(0, 2 * math.pi)
        self.velocity: p.Vector2 = p.Vector2(
            math.cos(angle) * speed,
            math.sin(angle) * speed,
        )
        self.color: p.Color = p.Color(
            random.randint(0, 255),
            random.randint(0, 255),
            random.randint(0, 255),
            255,
        )

    def update(self, dt: float) -> None:
        # move
        self.center.x += self.velocity.x * dt
        self.center.y += self.velocity.y * dt

        # set bounds
        if self.center.x < self.radius:
            self.center.x = self.radius
            self.velocity.x *= -1

        if self.center.x > p.get_screen_width() - self.radius:
            self.center.x = p.get_screen_width() - self.radius
            self.velocity.x *= -1

        if self.center.y < self.radius:
            self.center.y = self.radius
            self.velocity.y *= -1

        if self.center.y > p.get_screen_height() - self.radius:
            self.center.y = p.get_screen_height() - self.radius
            self.velocity.y *= -1

    def draw(self) -> None:
        p.draw_circle_v(self.center, self.radius, self.color)


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, "Bouncing Balls")

    balls: list[Ball] = [Ball() for _ in range(MAX_BALLS)]

    while not p.window_should_close():
        dt: float = p.get_frame_time()

        for ball in balls:
            ball.update(dt)

        p.begin_drawing()
        p.clear_background(p.RAYWHITE)

        for ball in balls:
            ball.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/single_file_programs/bouncing_balls_v2.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <random>
#include <raylib.h>

constexpr int SCREEN_WIDTH = 800;
constexpr int SCREEN_HEIGHT = 600;
constexpr int MAX_BALLS = 200;

static inline float random_float(const float min, const float max)
{
    if (min == max)
        return min;

    const float minimum = std::min(min, max);
    const float maximum = std::max(min, max);

    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<float> distr(minimum, maximum);

    return distr(gen);
}

class Ball {
  public:
    float radius;
    Vector2 center;
    Vector2 velocity;
    Color color;

    Ball()
        : radius(random_float(5, 30)),
          center({random_float(radius, GetScreenWidth() - radius),
                  random_float(radius, GetScreenHeight() - radius)}),
          color({static_cast<unsigned char>(random_float(0, 255)),
                 static_cast<unsigned char>(random_float(0, 255)),
                 static_cast<unsigned char>(random_float(0, 255)), 255})
    {
        const float speed = random_float(100, 300);
        const float random_angle = random_float(0, 2 * PI);

        velocity = {std::cos(random_angle) * speed,
                    std::sin(random_angle) * speed};
    }

    void draw(void) const { DrawCircleV(center, radius, color); }

    void update(const float dt)
    {
        // move
        center.x += velocity.x * dt;
        center.y += velocity.y * dt;

        // bounds
        if (center.x < radius || center.x > GetScreenWidth() - radius) {
            velocity.x *= -1;
        }
        if (center.y < radius || center.y > GetScreenHeight() - radius) {
            velocity.y *= -1;
        }
    }
};

int main(void)
{
    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, "Bouncing Balls");
    SetTargetFPS(60);

    std::array<Ball, MAX_BALLS> balls;

    while (!WindowShouldClose()) {

        const float dt = GetFrameTime();

        for (Ball &ball : balls) {
            ball.update(dt);
        }

        BeginDrawing();
        ClearBackground(RAYWHITE);

        for (const Ball &ball : balls) {
            ball.draw();
        }

        EndDrawing();
    }

    CloseWindow();

    return 0;
}
